### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by zxdcm**

### CODE REVIEW

File: BookingSystem/BookingSystem.Commands/Commands/HotelCommands/Commands/EditHotelCommand.cs [hunk 1]
```diff
@@ -28,17 +28,18 @@ public EditHotelCommandHandler(BookingWriteContext dataContext, IMapper mapper)
...
 
-        public async Task<Result> Execute(EditHotelCommand command)
+        public async Task<Result> ExecuteAsync(EditHotelCommand command)
         {
-            var dto = command.Hotel;
+            var hotelDto = command.Hotel;
 
-            var hotel = await _dataContext.Hotels.FindAsync(dto.HotelId);
+            var hotel = await _dataContext.Hotels.FindAsync(hotelDto.HotelId);
             if (hotel == null)
```
Comment (ssachkovskaya) [inline BookingSystem/BookingSystem.Commands/Commands/HotelCommands/Commands/EditHotelCommand.cs:36]: Why do you use isActive for hotel, but isAvailable for extra services. Are there different delete operations?

File: BookingSystem/BookingSystem.Commands/Commands/HotelCommands/Commands/EditHotelExtraServicesCommand.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+using AutoMapper;
+using BookingSystem.Commands.Commands.HotelCommands.DTOs;
+using BookingSystem.Commands.Infrastructure;
+using BookingSystem.Common.Interfaces;
+using BookingSystem.WritePersistence;
+using BookingSystem.WritePersistence.WriteModels;
+
+namespace BookingSystem.Commands.Commands.HotelCommands.Commands
+{
+    public class EditHotelExtraServicesCommand : ICommand<Result>
```
Comment (ssachkovskaya) [inline BookingSystem/BookingSystem.Commands/Commands/HotelCommands/Commands/EditHotelExtraServicesCommand.cs]: where are you going to use this command taking into account you already have add/edit/delete extra service commands?

File: BookingSystem/BookingSystem.Commands/Commands/HotelCommands/DTOs/NewHotelDto.cs [hunk 3]
```diff
@@ -4,7 +4,7 @@ public class NewHotelDto
     {
         public string Name { get; set; }
         public string Address { get; set; }
-        public int CountryId { get; set; }
+//        public int CountryId { get; set; }
```
Comment (ssachkovskaya) [inline BookingSystem/BookingSystem.Commands/Commands/HotelCommands/DTOs/NewHotelDto.cs]: remove?

File: BookingSystem/BookingSystem.Commands/Infrastructure/Result.cs [hunk 4]
```diff
@@ -1,44 +1,56 @@
...
 
-        public static Result Error(string errorMessage) 
+        private Result(int? value, string errorMessage, IEnumerable<string> errors)
+            : base(value, false, errorMessage, errors)
+        { }
+
+        public static Result Ok()
+            => new Result(null, true, null);
+
+        public static Result Ok(int id) => new Result(id, true, null);
```
Comment (ssachkovskaya) [inline BookingSystem/BookingSystem.Commands/Infrastructure/Result.cs]: this should be written in 2 lines according your code style

File: BookingSystem/BookingSystem.Queries/Queries/BookingQueries/Queries/BookingDetailsQuery.cs [hunk 5]
```diff
@@ -24,19 +21,28 @@ public BookingDetailsQuery(int bookingId)
...
         public Task<BookingView> Execute(BookingDetailsQuery query)
         {
-            return _dataContext.Bookings.Select(BookingView.Projection)
-                .FirstOrDefaultAsync(b => b.BookingId == query.BookingId);
+            return _dataContext.Bookings
+                .Select(b => new BookingView()
+                {
+                    BookingId = b.BookingId,
+                    RoomId = b.RoomId,
+                    Status = b.Status.ToString(),
```
Comment (ssachkovskaya) [inline BookingSystem/BookingSystem.Queries/Queries/BookingQueries/Queries/BookingDetailsQuery.cs]: Are you sure you need to get the status as a string?

File: BookingSystem/BookingSystem.Queries/Queries/CityQueries/Queries/ListCitiesQuery.cs [hunk 6]
```diff
@@ -42,7 +42,7 @@ public async Task<IEnumerable<CityView>> Execute(ListCitiesQuery query)
                     CityId = city.CityId,
                     CityName = city.Name,
                     CountryId = city.CountryId,
-                    CountryName = city.Country.Name,
+                    //CountryName = city.Country.Name,
```
Comment (ssachkovskaya) [inline BookingSystem/BookingSystem.Queries/Queries/CityQueries/Queries/ListCitiesQuery.cs]: remove

File: BookingSystem/BookingSystem.Queries/Queries/CityQueries/Queries/CityDetailsQuery.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    public class CityDetailsQueryHandler : IQueryHandler<CityDetailsQuery, Task<CityView>>
+    {
+        private readonly BookingReadContext _dataContext;
+
+        public CityDetailsQueryHandler(BookingReadContext dataContext)
+        {
+            _dataContext = dataContext;
+        }
+
+        public async Task<CityView> Execute(CityDetailsQuery query)
```
Comment (ssachkovskaya) [inline BookingSystem/BookingSystem.Queries/Queries/CityQueries/Queries/CityDetailsQuery.cs]: ExecuteAsync?

File: BookingSystem/BookingSystem.Queries/Queries/CountryQueries/Queries/CountryDetailsQuery.cs [hunk 8]
```diff
@@ -1,14 +1,11 @@
...
 using BookingSystem.Common.Interfaces;
-using BookingSystem.Queries.Views;
+using BookingSystem.Queries.Queries.CountryQueries.Views;
 using BookingSystem.ReadPersistence;
 using Microsoft.EntityFrameworkCore;
 
-namespace BookingSystem.Queries.CountryQueries
+namespace BookingSystem.Queries.Queries.CountryQueries.Queries
 {
     public class CountryDetailsQuery : IQuery<Task<CountryView>>
```
Comment (ssachkovskaya) [inline BookingSystem/BookingSystem.Queries/Queries/CountryQueries/Queries/CountryDetailsQuery.cs]: Can you please remind me what case will be when you need the single city/country details?

File: BookingSystem/BookingSystem.Queries/Queries/HotelQueries/Queries/HotelDetailsQuery.cs [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+                from service in services.DefaultIfEmpty()
+                select new HotelDetailsView()
+                {
+                    HotelId = hotel.HotelId,
+                    Name = hotel.Name,
+                    Address  = hotel.Address,
+                    CountryName = country.Name,
+                    CityName = city.Name,
+                    IsActive = hotel.IsActive,
+//                    ExtraServices = service.Select(se => new ExtraServiceView()
```
Comment (ssachkovskaya) [inline BookingSystem/BookingSystem.Queries/Queries/HotelQueries/Queries/HotelDetailsQuery.cs]: Is it still in work? If no, please, remove

File: BookingSystem/BookingSystem.ReadPersistence/ReadModels/Booking.cs [hunk 10]
```diff
@@ -1,26 +1,20 @@
...
-    public partial class Booking
+    public class Booking
     {
-        public Booking()
-        {
-            BookingExtraServices = new HashSet<BookingExtraService>();
-        }
+        public Booking() { }
 
         public int BookingId { get; set; }
```
Comment (ssachkovskaya) [inline BookingSystem/BookingSystem.ReadPersistence/ReadModels/Booking.cs]: Can investigate if properties can be with private set?

File: BookingSystem/BookingSystem.WebApi/Controllers/BookingController.cs [hunk 11]
```diff
@@ -49,18 +46,23 @@ public async Task<IActionResult> GetBooking(int id)
...
         // POST: api/Booking
         [HttpPost]
-        public async Task<IActionResult> Post([FromBody] NewBookingDto booking)
+        public async Task<IActionResult> AddBooking([FromBody] NewBookingDto booking)
         {
-            var result = await _commandDispatcher.Dispatch(new BookRoomCommand(booking));
-            return FromResult(result);
+            var result = await _commandDispatcher.DispatchAsync(new BookRoomCommand(booking));
+            if (result.IsSuccessful == false)
+                return BadRequest(result);
```
Comment (ssachkovskaya) [inline BookingSystem/BookingSystem.WebApi/Controllers/BookingController.cs]: BadRequest is usually used for wrong input. Think about other types of errors

File: BookingSystem/BookingSystem.Commands/Commands/AccountCommands/Commands/SignInCommand.cs [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        {
+            var userDto = command.User;
+
+            var user = await _dataContext.Users.FirstOrDefaultAsync(u => u.Email == userDto.Email);
+            if (user == null)
+                return Result.Error(string.Format(ErrorsResources.UserNotFound, userDto.Email));
+
+            if (_hasher.VerifyPassword(user.PasswordHash, userDto.Password))
+            {
+                return Result.Ok(user.UserId);
```
Comment (ssachkovskaya) [inline BookingSystem/BookingSystem.Commands/Commands/AccountCommands/Commands/SignInCommand.cs]: I would make Ok result the last one in this method in order not to interrupt the chain of error results

File: BookingSystem/BookingSystem.Commands/Commands/BookingCommands/Commands/BookExtraServicesCommand.cs [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        public async Task<Result> ExecuteAsync(BookExtraServicesCommand command)
+        {
+            var bookingDto = command.Booking;
+
+            var booking = await _dataContext.Bookings.FindAsync(bookingDto.BookingId);
+            if (booking == null)
+                return Result.NullEntityError(nameof(Booking), bookingDto.BookingId);
+
+            var result = await _bookingService.CanBookExtraServicesAsync(booking, bookingDto.ExtraServicesIds);
+            if (!result)
```
Comment (ssachkovskaya) [inline BookingSystem/BookingSystem.Commands/Commands/BookingCommands/Commands/BookExtraServicesCommand.cs]: Result can be any type, it is not clear that this result is boolean. Please make it clear with the variable name

File: BookingSystem/BookingSystem.Queries/Queries/HotelQueries/Queries/HotelDetailsQuery.cs [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+                    hotel.Name,
+                    hotel.Address,
+                    CountryName = country.Name,
+                    CityName = city.Name,
+                    hotel.IsActive,
+                }
+                into gr
+                select gr;
+            var hotelGr = await hotelViews.FirstOrDefaultAsync();
+            return new HotelDetailsView
```
Comment (ssachkovskaya) [inline BookingSystem/BookingSystem.Queries/Queries/HotelQueries/Queries/HotelDetailsQuery.cs:130]: isn't it possible to select HotelDetailsView as a part of query and then apply FirstOrDefaultAsync?

File: BookingSystem/BookingSystem.Commands/Commands/AccountCommands/Commands/SignUpCommand.cs [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+
+        public async Task<Result> ExecuteAsync(SignUpCommand command)
+        {
+            var userDto = command.User;
+
+            var userExist = await _dataContext.Users.AnyAsync(u => u.Email == userDto.Email);
+            if (userExist)
+                return Result.Error(string.Format(ErrorsResources.UserAlreadyExist, userDto.Email));
+
+            var user = _mapper.Map<User>(userDto);
```
Comment (ssachkovskaya) [inline BookingSystem/BookingSystem.Commands/Commands/AccountCommands/Commands/SignUpCommand.cs:50]: Are you planning to have any validation for empty email, too short password and etc.?

File: BookingSystem/BookingSystem.Commands/Commands/HotelCommands/Commands/EditHotelCommand.cs (hunk 1 above)
Comment (zxdcm) [reply BookingSystem/BookingSystem.Commands/Commands/HotelCommands/Commands/EditHotelCommand.cs:36]: Fixed.

**Pull request merged**